libs/sdk-ui-dashboard/src/model/queryServices/queryInsightDateDatasets.ts

```diff
@@ -0,0 +1,287 @@
+// (C) 2021 GoodData Corporation
+import { DashboardContext } from "../types/commonTypes";
+import { SagaIterator } from "redux-saga";
+import {
+    areObjRefsEqual,
+    filterObjRef,
+    idRef,
+    IInsight,
+    insightFilters,
+    insightRef,
+    isDateFilter,
+    ObjRef,
+    objRefToString,
+    serializeObjRef,
+} from "@gooddata/sdk-model";
+import { call, SagaReturnType, select } from "redux-saga/effects";
+import { createCachedQueryService } from "../state/_infra/queryService";
+import { PromiseFnReturnType } from "../types/sagas";
+import {
+    InsightAttributesMeta,
+    InsightDateDatasets,
+    queryInsightAttributesMeta,
+    QueryInsightDateDatasets,
+} from "../queries";
+import { selectObjectAvailabilityConfig } from "../state/config/configSelectors";
+import { selectInsightByRef } from "../state/insights/insightsSelectors";
+import { invalidQueryArguments } from "../events/general";
+import { ICatalogDateDataset, isCatalogDateDataset } from "@gooddata/sdk-backend-spi";
+import { query } from "../state/_infra/queryCall";
+import { invariant } from "ts-invariant";
+import { selectCatalogDateDatasets } from "../state/catalog/catalogSelectors";
+import uniqBy from "lodash/uniqBy";
+import fromPairs from "lodash/fromPairs";
+
+export const QueryDateDatasetsForInsightService = createCachedQueryService(
+    "GDC.DASH/QUERY.INSIGHT.DATE.DATASETS",
+    queryService,
+    (query: QueryInsightDateDatasets) => serializeObjRef(query.payload.insightRef),
+);
+
+/**
+ * Selector that will return date datasets for insight. The input to the selector is the dashboard query that is used
+ * to obtain and cache the data.
+ *
+ * This selector will return undefined if the query to obtain the data for particular insight was not yet fired or
+ * processed. Otherwise will return object containing `status` of the data retrieval; if the `status` is
+ * `'success'` then the `result` prop will contain the data.
+ *
+ * @remarks see {@link QueryInsightDateDatasets}
+ * @internal
+ */
+export const selectDateDatasetsForInsight = QueryDateDatasetsForInsightService.cache!.selectQueryResult;
```

Comment:
Can we somehow avoid the need for the `!.`?

Reply:
added `Required` to the typing of `createCachedQueryService` return type.